xtensa: intel_s1000: turn on XTENSA_ASM2

dcpleung opened the pull request:
This adds the necessary bits to support ASM2 with XCC for Intel S1000 SoC. With ASM2 enabled, xcc is now required to build for Intel S1000 SoC.

This also fixes some trivial compilation errors/warnings.

andyross commented on drivers/timer/xtensa_sys_timer.c:
```diff
@@ -433,11 +433,16 @@ void _zxt_tick_timer_init(void)
+	 * Since CCOMPARE* registers have undefined values after reset,
+	 * set compare value first beforing enabling timer interrupt.
+	 */
+	SET_TIMER_FIRE_TIME(GET_TIMER_CURRENT_TIME() + _xt_tick_divisor);
```
Can you yank this one out of the PR?  I'm finishing up a completely new xtensa timer driver and am too lazy to rebase. 

I promise it will initialize the CCOMPARE register correction. :)

dcpleung replied:
Done.

andyross commented on soc/xtensa/intel_s1000/include/_soc_inthandlers.h:
```diff
@@ -0,0 +1,257 @@
+ * dispatcher will handle exactly one flagged interrupt, in numerical
+ * order (low bits first) and will return a mask of that bit that can
+ * then be cleared by the calling code.  Unrecognized bits for the
+ * level will invoke an error handler.
```
This reminds me: I didn't have enough cmake-fu to get the python script to generate these from the core-isa.h at build time, which is why these files are checked into git.  I bet I could do it now...

andyross commented on boards/xtensa/intel_s1000_crb/board.cmake:
```diff
@@ -7,3 +7,7 @@ board_finalize_runner_args(intel_s1000
 )
+
+if(NOT "${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "xcc")
+  message(FATAL_ERROR "ZEPHYR_TOOLCHAIN_VARIANT != xcc. This requires xcc to build!")
```
Out of curiosity: what is it that actually requires xcc?  I mean... I understand that no one has built a gcc toolchain for S1000 and that no one plans to, but AFAIK there are no xcc extensions in use and in theory a perfectly valid app could be generated with a gcc toolchain.  Seems like you could take this out.

dcpleung replied:
Without XCC, the #error in _soc_inthandlers.h are triggered. This is because of all the missing XCHAL_* from GCC.

nashif replied:
AFAIK it was never possible to build this board with anything but XCC for the S1000.

andyross replied:
Yeah, on Xtensa you have to have a per-SoC toolchain built from their "overlay" output that comes from the synthesis tools.  And no one has done that.  My point was that in principle this could be done, and that the platform could then be buildable with gcc.  We aren't using any xcc-specific stuff.  Pedantry, basically.

nashif replied:
AFAIK @@lgirdwood had a gcc toolchain based on xtools for the S1000, will check with him

andyross reviewed (changes requested):
The code all looks great.  I'm giving it a -1 out of my purely parochial interest not to have to rebase an in-flight timer driver change. :)

Feel free to override if this needs to go in sooner.

dcpleung committed:
audio: tlv320dac310x.c:141: fix too few arguments for format

The logging statement specifies a parameter but one was never
provided. So add it.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: specify which SR to store pointer to _kernel.cpu struct

This allows Kconfig to specify which special register is being
used to store the pointer to the _kernel.cpu struct.
Since the SoC itself is highly configurable, sometimes MISC0 is not
available. So this adds the ability to use other special registers.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
dts: intel_s1000: fix warning about leading "0x"

This removes the prefix "0x" for sram0 to fix a warning.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
toolchain: xcc: fake __builtin_umul[ll]_overflow

XCC doesn't provide these builtins so we have to define them
with minimal functionality for testing.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: intel_s1000: turn on XTENSA_ASM2

This adds the necessary bits to support ASM2 with XCC
for Intel S1000 SoC. With ASM2 enabled, xcc is now
required to build for Intel S1000 SoC.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andyross reviewed (approved):
Remove -1.   Everything looks clean to me.

lgirdwood commented:
@dcpleung I don't see anything in this PR that would block gcc in the long term (unless this is enabling some dormant code that uses HiFi DSP instructions ?). Adding new registers for rsr/wsr is fairly trivial and you may find them already added in gcc 8.1 used by SOF.

https://github.com/thesofproject/xtensa-overlay

dcpleung commented:
@lgirdwood I tried building with XTENSA_ASM2 enabled with GCC (ignoring all the errors in ```_soc_inthandlers.h```), and it built fine. However, there was a double-fault after ```switch_to_main_thread()```.

andyross commented:
But you were building with the GCC in our SDK, right?  The SDK is based on the qemu CPU, which (among other differences) has half as many registers in the window set and would indeed be expected to explode on context switch. :)

My point was just that a toolchain built from a proper overlay (i.e. the one linked above) "should" work, and that we probably shouldn't be hard-configuring "xcc".   Totally a minor point.

But if we do have such a toolchain built, I'd be really curious to see the results.

dcpleung commented:
Yeah, once we have a proper GCC toolchain, we can lift the XCC restriction.

lgirdwood commented:
No I'm using SOF gcc to build SOF for CNL target (same ISA as s1000). You may want to try this version of gcc as it may help. 
Fwiw adding new rsr and wsr target regs for qemu are also fairly straight forward mechanical changes if you have the time.
